src/graph/GraphViewsManager.ts

```diff
@@ -69,41 +104,180 @@ export class GraphViewsManager implements IServerProvider {
     this._servers.set(id, server);
     return id;
   }
+
+  public findPanelById(id: number): vscode.WebviewPanel {
+    return this._panels.get(id);
+  }
+
 }
 
-class GraphViewDocumentContentProvider implements vscode.TextDocumentContentProvider {
+class GraphViewDocumentContentProvider {
   public onDidChange?: vscode.Event<vscode.Uri>;
 
-  public constructor(private _serverProvider: IServerProvider) { }
+  public constructor(private _serverProvider: IServerProvider, private _context: vscode.ExtensionContext) { }
 
-  public provideTextDocumentContent(uri: vscode.Uri, _token: vscode.CancellationToken): vscode.ProviderResult<string> {
+  public provideHtmlContent(uri: vscode.Uri): string {
+    // the uri has format: "1", "2" ...
     // Figure out which client to attach this to
     // tslint:disable-next-line:no-single-line-block-comment
-    let serverId = parseInt(uri.path.slice(1) /* remove '/' from beginning */, 10);
+    let serverId = parseInt(uri.path, 10);
     console.assert(serverId > 0);
     let server = this._serverProvider.findServerById(serverId);
     if (server) {
-      let clientHtmlPath = path.join(resourcesPath, 'graphClient', 'graphClient.html');
-      console.assert(fs.existsSync(clientHtmlPath), `Couldn't find ${clientHtmlPath}`);
+      let options: IGraphClientHtmlPaths = {
+        d3Uri: this._getVscodeResourceUri(['dist', 'node_modules', 'd3', 'd3.js']),
+        socketUri: this._getVscodeResourceUri(['dist', 'node_modules', 'socket.io-client', 'dist', 'socket.io.js']),
+        graphClientUri: this._getVscodeResourceUri(['dist', 'graphClient.js']),
+        cosmosIconUri: this._getVscodeResourceUri(['resources', 'cosmos.png']),
+        cssUri: this._getVscodeResourceUri(['resources', 'graphClient', 'graphClient.css'])
+      };
+      return this._graphClientHtmlAsString(server.port, options);
+    }
+
+    return "This resource is no longer available.";
+  }
+
+  private _getVscodeResourceUri(direcotryList: string[]): string {
+    let uri = vscode.Uri.parse(path.join("file:" + this._context.extensionPath, ...direcotryList));
+    return `vscode-resource:${uri.fsPath}`;
+  }
 
-      let html = `
+  private _graphClientHtmlAsString(port: number, options: IGraphClientHtmlPaths): string {
+    return `
     <!DOCTYPE html>
     <html>
-      <style>
-        body {
-          padding: 0;
-          margin: 0;
-        }
-      </style>
-      <body>
-        <iframe src="file://${clientHtmlPath}?port=${server.port}" style="width: 100%; height: 100%; position: absolute; padding: 0; margin: 0; border: none"></iframe>
-      </body>
+
```

Comment:
I assume you'll move this back into the separate HTML file...

Reply:
I agree. I found it harder to edit myself. How would the `options.uri` strings work once the HTML is read from disk though?